pkg/reconciler/autoscaling/kpa/scaler.go

```diff
@@ -371,6 +376,38 @@ func (ks *scaler) scale(ctx context.Context, pa *autoscalingv1alpha1.PodAutoscal
 		return desiredScale, nil
 	}
 
+	// Before we apply scale to zero and since we have failed to activate, check if any pod is in waiting state
+	// This should capture any case where we scale from zero and regular progressDeadline expiration will not be hit due to K8s limitations
+	// when not being in a deployment rollout.
+	if desiredScale == 0 && pa.Status.IsActivating() && podForErrorChecking != nil {
+		if err = checkAvailabilityBeforeScalingDown(ctx, logger, podForErrorChecking, pa, client); err != nil {
+			return desiredScale, fmt.Errorf("failed to get check availability for target %v: %w", pa.Spec.ScaleTargetRef, err)
+		}
+	}
+
 	logger.Infof("Scaling from %d to %d", currentScale, desiredScale)
 	return desiredScale, ks.applyScale(ctx, pa, desiredScale, ps)
 }
+
+func checkAvailabilityBeforeScalingDown(ctx context.Context, logger *zap.SugaredLogger, pod *corev1.Pod, pa *autoscalingv1alpha1.PodAutoscaler, client clientset.Interface) error {
+	for _, status := range pod.Status.ContainerStatuses {
+		if status.Name != revresurces.QueueContainerName {
+			if w := status.State.Waiting; w != nil {
+				logger.Debugf("marking revision inactive with: %s: %s", w.Reason, w.Message)
+				pa.Status.MarkScaleTargetNotInitialized(w.Reason, w.Message)
+				return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+					rev, err := client.ServingV1().Revisions(pa.Namespace).Get(ctx, pa.Name, metav1.GetOptions{})
+					if err != nil {
+						return err
+					}
+					rev.Status.MarkResourcesAvailableFalse(w.Reason, w.Message)
+					if _, err = client.ServingV1().Revisions(pa.Namespace).UpdateStatus(ctx, rev, metav1.UpdateOptions{}); err != nil {
+						return err
+					}
+					return nil
+				})
+			}
+		}
```

Comment:
this is sorta violating our abstractions - if we wanted to propagate this error message to the revision we would have to do it through a PodAutoscaler condition.

Reply:
> have to do it through a PodAutoscaler condition.

Normally yes but due to the distributed status logic (not clear, undocumented) and how things are implemented this is safer imho. I can try change but it will require to propagate this decision down to the pa status update (not ideal), I had it this ways previously.